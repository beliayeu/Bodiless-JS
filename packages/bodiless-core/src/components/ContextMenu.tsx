/**
 * Copyright Â© 2019 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { FC, useState } from 'react';
import { getUI as getFormUI } from '../contextMenuForm';
import {
  IContextMenuProps as IProps, UI, ContextMenuFormProps, TMenuOption,
} from '../Types/ContextMenuTypes';
import ContextMenuItem from './ContextMenuItem';
import StructuredChildren from '../ContextMenu/StructuredChildren';

const defaultUI = {
  Toolbar: 'div',
};

export const getUI = (ui: UI = {}) => ({
  ...defaultUI,
  ...getFormUI(),
  ...ui,
});

const createChildrenFromOptions = (options: TMenuOption[]) => options.map(
  option => {
    const Component = option.Component || ContextMenuItem;
    return (
      <Component
        option={option}
        group={option.group || option.name}
        name={option.name}
        key={option.name}
        aria-label={option.name}
      />
    );
  },
);

// @TODO: This should come from the ui.
const Group: FC<any> = ({ children }) => <>{children}</>;

const ContextMenu: FC<IProps> = (props) => {
  if (typeof window === 'undefined') return null;

  const [renderForm, setRenderForm] = useState<(props:ContextMenuFormProps) => JSX.Element>();
  const {
    options,
    ui,
    renderInTooltip = true,
    children,
  } = props;
  const { Toolbar } = getUI(ui);

  const childProps = {
    ui,
    setParentRenderForm: renderInTooltip ? undefined : setRenderForm,
  };
  const childrenFromOptions = createChildrenFromOptions(options);

  if (renderForm) {
    const formProps: ContextMenuFormProps = {
      closeForm: () => setRenderForm(undefined),
      ui,
      'aria-label': 'Context Submenu Form',
    };
    return renderForm(formProps);
  }
  const finalChildren = children
    ? [...React.Children.toArray(children).filter(React.isValidElement), ...childrenFromOptions]
    : childrenFromOptions;

  if (finalChildren.length > 0) {
    return (
      <Toolbar onClick={(e: React.MouseEvent<HTMLDivElement>) => e.stopPropagation()}>
        <StructuredChildren components={{ Group }} {...childProps}>
          {finalChildren}
        </StructuredChildren>
      </Toolbar>
    );
  }

  return null;
};

export default ContextMenu;
