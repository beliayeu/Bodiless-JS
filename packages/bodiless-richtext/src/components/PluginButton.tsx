/**
 * Copyright Â© 2020 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { ComponentType } from 'react';
import MaterialIcon from '@material/react-material-icon';
import { useUI } from '../RichTextContext';

const DefaultPluginButton = props  => <button {...props} />

type uiIndexType = {
  [index: string]: any;
};

type Props = {
  componentName?: string;
  Component?: ComponentType<any>;
  icon: string;
  // ToDo improve types
  children: any;
}

const PluginButton = (props: Props) => {
  const { componentName, Component, icon, children, ...rest } = props;

  // Workaround to get styled component from UI if it exists with fallback to original Component.
  const completeUI: uiIndexType = useUI();
  let StyledComponent;
  if (componentName && completeUI[componentName]) {
    StyledComponent = completeUI[componentName];
  } else if (Component) {
    StyledComponent = Component;
  } else {
    StyledComponent = DefaultPluginButton;
  }

  return (
    <StyledComponent {...rest}>
      {children || <MaterialIcon className="bl-material-icons" icon={icon} />}
    </StyledComponent>
  );
}

export default PluginButton;
