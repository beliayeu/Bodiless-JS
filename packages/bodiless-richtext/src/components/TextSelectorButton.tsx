/**
 * Copyright Â© 2019 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useState } from 'react';
import RCTooltip from 'rc-tooltip';
import MaterialIcon from '@material/react-material-icon';
import { useUI, getUI } from '../RichTextContext';

type ButtonProps = {
  className: string
};

const NodeSelectorButton = (props: ButtonProps) => {
  const { Button } = getUI(useUI());

  return (
    <Button {...props}>
      <MaterialIcon className="bl-material-icons" icon="more_horiz" />
    </Button>
  );
};

NodeSelectorButton.displayName = 'NodeSelectorButton';

type props = {
  children: React.ReactNode,
};

const TextSelectorButton = ({
  children,
}:props) => {
  const [visible, setVisible] = useState(false);
  const { Overlay, TextSelectorWrapper } = getUI(useUI());
  const nodeSelectorProps = {
    className: visible ? 'bl-active node-selector-button' : '',
  };

  return (
    <RCTooltip
      trigger={['hover']}
      placement="topLeft"
      overlayStyle={{ opacity: 1 }}
      visible={visible}
      align={{
        offset: [-40, -10],
      }}
      onVisibleChange={() => { setVisible(!visible); }}
      overlay={() => (
        <Overlay>
          <TextSelectorWrapper
            onClick={(ev) => {
              const target = ev.target as HTMLElement;
              const currentTarget = ev.currentTarget as HTMLElement;
              if (currentTarget !== target && currentTarget.contains(target)) {
                const buttons = document.getElementsByClassName('node-selector-button');
                if (buttons.length) {
                  const button = buttons[0] as HTMLElement;
                  button.focus();
                }
                setVisible(!visible);
              }
            }}
          >
            { children }
          </TextSelectorWrapper>
        </Overlay>
      )}
    >
      <NodeSelectorButton {...nodeSelectorProps} />
    </RCTooltip>
  );
};

TextSelectorButton.displayName = 'TextSelectorButton';

export default TextSelectorButton;
