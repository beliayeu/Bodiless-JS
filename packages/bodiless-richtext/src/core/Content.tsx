/**
 * Copyright Â© 2019 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { Editable, DefaultElement, DefaultLeaf } from 'slate-react';
import { useSlateContext } from './SlateEditorContext';
import type { EditorContext } from '../Type';

type EditableProps = EditorContext['editorProps'];

const withWrapper = WrapperComponent => Component => ({children, ...rest}) => (
  <WrapperComponent {...rest}>
    <Component>{children}</Component>
  </WrapperComponent>
)


const renderLeaf = (props) => {
  const { leaf } = props;
  const editorContext = useSlateContext();
  const plugins = editorContext?.editorProps!.plugins;
  let renderLeaf$ = DefaultLeaf;
  plugins.forEach(plugin => {
    if (plugin.hasOwnProperty('renderLeaf') && leaf[plugin.type]) {
      renderLeaf$ = withWrapper(plugin.renderLeaf)(renderLeaf$);
    }
  });
  return renderLeaf$(props);
}

const renderElement = (props) => {
  const { element } = props;
  const editorContext = useSlateContext();
  const plugins = editorContext?.editorProps!.plugins;
  let renderElement$ = DefaultElement;
  plugins.forEach(plugin => {
    if (plugin.hasOwnProperty('renderElement') && element.type === plugin.type) {
      renderElement$ = plugin.renderElement;
    }
  });
  return renderElement$(props);
}

const Content = (props: EditableProps) => {
  const editorContext = useSlateContext();
  return (
    <Editable
      {...props}
      { ...editorContext!.editorProps}
      renderLeaf={renderLeaf}
      renderElement={renderElement}
    />
  );
};

export default Content;
