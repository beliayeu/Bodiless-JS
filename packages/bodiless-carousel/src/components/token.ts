/**
 * Copyright Â© 2021 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  withDesign,
  addProps,
  replaceWith,
  stylable,
  Ul,
  Li,
} from '@bodiless/fclasses';
import { ifEditable, withChild } from '@bodiless/core';
import type { WithNodeKeyProps } from '@bodiless/core';
import { asBodilessList } from '@bodiless/components';
import flow from 'lodash/flow';
import {
  ButtonBack,
  ButtonNext,
  ButtonPlay,
} from 'pure-react-carousel';
import CarouselDot from './CarouselDot';

const withIntrinsicHeight = withDesign({
  Wrapper: addProps({
    isIntrinsicHeight: true,
  }),
});

const withNoDragIfEditable = withDesign({
  Wrapper: ifEditable(
    addProps({
      dragEnabled: false,
    }),
  ),
});

const withInfinitiveLoop = withDesign({
  Wrapper: addProps({
    infinite: true,
  }),
});

const withCarouselDots = (nodeKeys?: WithNodeKeyProps) => flow(
  withDesign({
    Dots: flow(
      replaceWith(Ul),
      asBodilessList(nodeKeys),
      withDesign({
        Item: replaceWith(
          withChild(
            stylable(CarouselDot), 'Dot',
          )(Li),
        ),
      }),
    ),
  }),
);

const withNavigationButtons = withDesign({
  ButtonBack: replaceWith(stylable(ButtonBack)),
  ButtonNext: replaceWith(stylable(ButtonNext)),
});

const withAutoPlayButton = withDesign({
  ButtonPlay: replaceWith(stylable(ButtonPlay)),
});

const withAutoPlayInterval = (interval: number = 3000) => withDesign({
  Wrapper: addProps({
    interval,
  }),
});

export {
  withIntrinsicHeight,
  withNoDragIfEditable,
  withInfinitiveLoop,
  withCarouselDots,
  withNavigationButtons,
  withAutoPlayButton,
  withAutoPlayInterval,
};
