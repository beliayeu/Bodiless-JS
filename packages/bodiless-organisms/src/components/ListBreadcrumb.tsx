/**
 * Copyright Â© 2020 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, {
  useState,
  useEffect,
  ComponentType,
} from 'react';
import { flow, identity } from 'lodash';

import {
  withDesign,
  replaceWith,
  asComponent,
  withOnlyProps,
  extendDesignable,
  designable,
  Li,
  Span,
} from '@bodiless/fclasses';
import {  withSidecarNodes, WithNodeKeyProps } from '@bodiless/core';
import {
  asBreadcrumb,
  useIsActiveBreadcrumb,
  withSimpleSubListDesign,
  useBreadcrumbContext,
} from '@bodiless/components';

type Settings = {
  nodeKeys?: WithNodeKeyProps;
  depth?: number;
}

/**
 * HOC which is used to hide a list item with inactive breadcrumb
 * wait until component is mounted, so that children are rendered
 * hide the component if breadcrumb is not active
 * @param Component 
 */
const withHiddenInactiveBreadcrumbsItem = (Component: ComponentType<any>) => {
  const WithHiddenInactiveBreadcrumbsItem = (props: any) => {
    const [ mounted, setMounted ] = useState(false);
    const isActive = useIsActiveBreadcrumb();
    useEffect(() => {
      setMounted(true);
    }, []);
    const props$1 = {
      ...props,
      className: !isActive || !mounted ? 'hidden' : '',
    }
    return !isActive && mounted ? <></> : (<Component {...props$1} />)
  }
  WithHiddenInactiveBreadcrumbsItem.displayName = 'WithHiddenInactiveBreadcrumbsItem';
  return WithHiddenInactiveBreadcrumbsItem;
}

const asBreadcrumbListItem = (nodeKeys?: WithNodeKeyProps) => withSidecarNodes(
  asBreadcrumb(nodeKeys),
  withHiddenInactiveBreadcrumbsItem,
);

const asBreadcrumbSubList = withDesign({
  Wrapper: withDesign({
    List: replaceWith(withOnlyProps('key', 'children')(React.Fragment)),
    WrapperItem: replaceWith(withOnlyProps('key', 'children')(React.Fragment)),
    SubListTitleWrapper: replaceWith((props: any) => {
      const [ mounted, setMounted ] = useState(false);
      useEffect(() => {
        setMounted(true);
      }, []);
      const props$1 = {
        ...props,
        className: !mounted ? 'hidden' : '',
      }
      return <Li {...props$1} />;
    }),
  }),
});

/**
 * HOC which can be applied to a base list to make it into a site's breadcrumbs
 *
 * @param A base list component created via asBodilessList()
 *
 * @return A clean (unstyled) site breadcrumb component.
 */
const asBreadcrumbsClean = ({ nodeKeys = 'link', depth = 1 }: Settings) => flow(
  /*withDesign({
    Item: flow(
      withBreadcrumbSeparator,
      withDesign({
        Separator: replaceWith(() => <div id="parentSeparator" />),
      }),
      withDesign({
        Item: flow(
          withBreadcrumbSeparator,
          withDesign({
            Separator: replaceWith(() => <div id="parentSeparator2" />),
          }),
        ),
      }),
    ),
  }),*/
  withSimpleSubListDesign(depth)(flow(
    asBreadcrumbListItem(nodeKeys),
    asBreadcrumbSubList,
  )),
);


// ToDo: remediate types
const withBreadcrumbSeparator = (Component: ComponentType<any>) => {
  const WithBreadcrumbSeparator = (props: any) => {
    const { components, ...rest } = props;
    const { Separator } = components;
    const isActiveBreadcrumb = useIsActiveBreadcrumb();
    return (
      <>
        <Component {...rest} />
        { isActiveBreadcrumb &&  <Separator /> }
      </>
    );
  }
  WithBreadcrumbSeparator.displayName = 'WithBreadcrumbSeparator';
  const extendDesignable$ = extendDesignable(({Separator, ...rest}: any) => rest);
  return extendDesignable$({ Separator: asComponent('span') })(WithBreadcrumbSeparator);
}

const TrailItem = (props: any) => {
  const { components } = props;
  const { Wrapper, Item } = components;
  return (
    <Wrapper>
      <Item />
    </Wrapper>
  )
}
const DesignableTrailItem = designable({
  Wrapper: Li,
  Item: Span,
})(TrailItem);


// ToDo: remediate types
const withBreadcrumbCustomItems = (Component: ComponentType<any>) => {
  const WithBreadcrumbCustomItems = (props: any) => {
    const { components, children, ...rest } = props;
    const { StartingTrail, Separator, FinalTrail, } = components;
    const breadcrumbContext = useBreadcrumbContext();
    console.log('hey from withBreadcrumbCustomItems');
    console.log(breadcrumbContext.hasActive)
    return (
      <Component {...rest}>
        <StartingTrail />
        { StartingTrail && <Separator /> }
        {children}
        { FinalTrail && !breadcrumbContext.hasActive  && <Separator /> }
        { !breadcrumbContext.hasActive && <FinalTrail /> }
      </Component>
    );
  }
  WithBreadcrumbCustomItems.displayName = 'withBreadcrumbCustomItems';
  const extendDesignable$ = extendDesignable(({
    StartingTrail, Separator, FinalTrail, ...rest
  }: any) => rest);
  return extendDesignable$({
    StartingTrail: DesignableTrailItem,
    Separator: asComponent('span'),
    FinalTrail: DesignableTrailItem,
  })(WithBreadcrumbCustomItems);
}

/*const withBreadcrumbSeparatorDesign = withSimpleSubListDesign(2)(withDesign({
  Item: flowRight(
    withBreadcrumbSeparator
  ),
}))*/

const defaultDepth = 3;

const withBreadcrumbDesign = ({ Title: TitleDesign, Item: ItemDesign, Separator: SeparatorDesign }: any) => flow(
  TitleDesign ? flow(
    // @ts-ignore ToDo: resolve types
    withSimpleSubListDesign(defaultDepth)(withDesign({ Title: TitleDesign })),
    withDesign({ Title: TitleDesign }),
  ) : identity,
  ItemDesign ?
    withSimpleSubListDesign(defaultDepth)(flow(
      withDesign({ Item: ItemDesign }),
      withDesign({
        Wrapper: withDesign({
          SubListTitleWrapper: ItemDesign,
        }),
      }),
    ),
  ) : identity,
);

export default asBreadcrumbsClean;
export {
  withHiddenInactiveBreadcrumbsItem,
  withBreadcrumbSeparator,
  asBreadcrumbListItem,
  asBreadcrumbSubList,
  withBreadcrumbDesign,
  withBreadcrumbCustomItems,
};


