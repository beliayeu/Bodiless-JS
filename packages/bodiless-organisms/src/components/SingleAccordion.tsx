/**
 * Copyright Â© 2019 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, {
  ComponentType,
  useState,
  createContext,
  useContext,
} from 'react';
import { flow } from 'lodash';
import {
  withDesign,
  designable,
  Div,
  H2,
  StylableProps,
  addProps,
} from '@bodiless/fclasses';
import type {
  DesignableComponentsProps
} from '@bodiless/fclasses';
import {
  asEditable,
} from '@bodiless/components';
import { withNode } from '@bodiless/core';

export type SingleAccordionComponents = {
  Wrapper: ComponentType<StylableProps>,
  TitleWrapper: ComponentType<StylableProps>,
  Title: ComponentType<StylableProps>,
  BodyWrapper: ComponentType<StylableProps>,
  Body: ComponentType<StylableProps>,
};
const singleAccordionComponentStart:SingleAccordionComponents = {
  Wrapper: Div,
  TitleWrapper: Div,
  Title: H2,
  BodyWrapper: Div,
  Body: Div,
};

enum ACCORDION_STATE {
  EXPANDED,
  COLLAPSED,
}

type Props = {
  state: ACCORDION_STATE,
  toggle: Function,
  expandedStyle: string | null,
} & DesignableComponentsProps<SingleAccordionComponents>;

const SingleAccordionBase = ({
  components, state, expandedStyle = null, toggle, ...rest
}: Props) => {
  const {
    Wrapper,
    TitleWrapper,
    Title,
    BodyWrapper,
    Body,
  } = components;

  return (
    <Wrapper className={[state]} {...rest}>
      <TitleWrapper
        onClick={toggle}
        className={[
          'flex',
          'justify-between',
          'select-none',
          (state === ACCORDION_STATE.EXPANDED && expandedStyle) || '',
        ]}
      >
        <Title />
        <span className="material-icons cursor-pointer select-none" data-accordion-element="accordion-icon" data-accordion-icon={state === ACCORDION_STATE.COLLAPSED ? 'expand' : 'collapse'}>
          {state === ACCORDION_STATE.COLLAPSED ? 'add' : 'remove'}
        </span>
      </TitleWrapper>
      <BodyWrapper className={state === ACCORDION_STATE.COLLAPSED ? 'hidden' : 'block'}>
        <Body />
      </BodyWrapper>
    </Wrapper>
  );
};

const createSingleAccordionProvider = () => {
  const SingleAccordionContext = createContext({
    accordionState: ACCORDION_STATE.EXPANDED,
    setAccordionState: () => {},
  });
  const withSingleAccordionProvider = Component => {
    const WithSingleAccordionProvider = props => {
      const { expanded } = props;
      const initialState = expanded ? ACCORDION_STATE.EXPANDED : ACCORDION_STATE.COLLAPSED;
      const [accordionState, setAccordionState] = useState(initialState);
      const providerValue = {
        accordionState,
        setAccordionState,
      };
      const props$1 = {
        ...props,
        state: accordionState,
        toggle: () => setAccordionState(accordionState === ACCORDION_STATE.EXPANDED ? ACCORDION_STATE.COLLAPSED : ACCORDION_STATE.EXPANDED),
      };
      return (
        <SingleAccordionContext.Provider value={providerValue}>
          <Component {...props$1} />
        </SingleAccordionContext.Provider>
      );
    };
    WithSingleAccordionProvider.displayName = 'WithSingleAccordionProvider';
    return WithSingleAccordionProvider;
  };
  return {
    withAccordionContext: withSingleAccordionProvider,
    useAccordionContext: () => useContext(SingleAccordionContext),
  };
};

const createSingleAccordion = () => {
  const { withAccordionContext, useAccordionContext } = createSingleAccordionProvider();
  return {
    Accordion: flow(
      designable(singleAccordionComponentStart),
      withAccordionContext,
    )(SingleAccordionBase),
    isExpanded: () => useAccordionContext().accordionState === ACCORDION_STATE.EXPANDED,
    isCollapsed: () => useAccordionContext().accordionState === ACCORDION_STATE.COLLAPSED,
  };
};

const asSingleAccordion = withDesign({
  Title: asEditable('title', 'SingleAccordion Title Text'),
  Body: asEditable('body', 'SingleAccordion Body Text'),
});

const asTestableAccordion = withDesign({
  Wrapper: addProps({ 'data-accordion-element': 'accordion' }),
  TitleWrapper: addProps({ 'data-accordion-element': 'accordion-title-wrapper' }),
  Title: addProps({ 'data-accordion-element': 'accordion-title' }),
  BodyWrapper: addProps({ 'data-accordion-element': 'accordion-body-wrapper' }),
  Body: addProps({ 'data-accordion-element': 'accordion-body' }),
});

const { Accordion: SingleAccordionClean } = createSingleAccordion();

const SingleAccordion = flow(
  asSingleAccordion,
  withNode,
)(SingleAccordionClean);

const TestableSingleAccordion = flow(
  asTestableAccordion,
  withNode,
)(SingleAccordionClean);

export default SingleAccordion;
export {
  createSingleAccordion,
  SingleAccordionBase,
  SingleAccordion,
  SingleAccordionClean,
  TestableSingleAccordion,
  asSingleAccordion,
  asTestableAccordion,
};
