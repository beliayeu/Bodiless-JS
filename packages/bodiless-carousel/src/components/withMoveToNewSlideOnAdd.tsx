/**
 * Copyright Â© 2021 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { ComponentType } from 'react';
import { useListContext, ListContext } from '@bodiless/components';
import { withDesign, withFinalDesign } from '@bodiless/fclasses';
import { useCarouselSetState } from './hooks';
import useBodilessListItemIndex from './useBodilessListItemIndex';

const withMoveToNewSlideOnAdd$ = (Component: ComponentType) => (props: any) => {
  const listContext = useListContext();
  const { addItem } = listContext;
  const carouselItemIndex = useBodilessListItemIndex();
  const setCarouselState = useCarouselSetState();
  const listContext$ = {
    ...listContext,
    addItem: () => {
      if (addItem !== undefined) addItem();
      setCarouselState({ currentSlide: carouselItemIndex + 1 });
    },
  };
  return (
    <ListContext.Provider value={listContext$}>
      <Component {...props} />
    </ListContext.Provider>
  );
};

const withMoveToNewSlideOnAdd = withDesign({
  Slider: withFinalDesign({
    Item: withMoveToNewSlideOnAdd$,
  }),
});

export default withMoveToNewSlideOnAdd;
