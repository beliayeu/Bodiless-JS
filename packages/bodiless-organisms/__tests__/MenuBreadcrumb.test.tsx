/**
 * Copyright Â© 2020 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { mount } from 'enzyme';
import { withDefaultContent, withSidecarNodes } from '@bodiless/core';
import { asBodilessLink } from '@bodiless/components';
import { replaceWith, withDesign, withOnlyProps } from '@bodiless/fclasses';
import { flowRight } from 'lodash';

import {
  asMenuBase,
  asBreadcrumbsClean,
} from '../src/components/Menu/SimpleMenu';

const { DefaultContentNode } = require('@bodiless/core');

const setPagePath = (pagePath: string) => {
  Object.defineProperty(DefaultContentNode.prototype, 'pagePath', {
    value: pagePath,
    writable: false,
  });
};

const createBreadcrumbComponent = ({
  content = {},
}) => flowRight(
  withDefaultContent(content),
  withDesign({
    // @ts-ignore
    BreadcrumbLink: replaceWith(withSidecarNodes(
      asBodilessLink(),
    ))('a'),
    BreadcrumbTitle: replaceWith(withOnlyProps('key', 'children')(React.Fragment)),
  }),
  asBreadcrumbsClean({
    linkNodeKey: 'title$link',
    titleNodeKey: 'title$text',
  }),
  asMenuBase('testMenu'),
)('ul');

describe('asBreadcrumbsClean', () => {
  it('creates breadcrumbs for basic 1-level menu', () => {
    setPagePath('/products');
    const Breadcrumb = createBreadcrumbComponent({
      content: {
        testMenu: {
          items: [
            'home',
            'products',
            'articles',
          ],
        },
        testMenu$home$title$link: {
          href: '/',
        },
        testMenu$products$title$link: {
          href: '/products',
        },
        testMenu$articles$title$link: {
          href: '/articles',
        },
      },
    });
    const wrapper = mount(<Breadcrumb />);
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('creates breadcrumbs for basic 2-level menu', () => {
    setPagePath('/products/productA');
    const Breadcrumb = createBreadcrumbComponent({
      content: {
        testMenu: {
          items: [
            'home',
            'products',
            'articles',
          ],
        },
        testMenu$home$title$link: {
          href: '/',
        },
        testMenu$products$title$link: {
          href: '/products',
        },
        testMenu$products$sublist: {
          items: [
            'productA',
            'productB',
          ],
        },
        'testMenu$products$cham-sublist': {
          component: 'SubMenu',
        },
        testMenu$products$sublist$productA$title$link: {
          href: '/products/productA',
        },
        testMenu$products$sublist$productB$title$link: {
          href: '/products/productB',
        },
        testMenu$articles$title$link: {
          href: '/articles',
        },
        testMenu$articles$sublist: {
          items: [
            'articleA',
          ],
        },
        testMenu$articles$sublist$articleA$title$link: {
          href: '/articles/articleA',
        },
      },
    });
    const wrapper = mount(<Breadcrumb />);
    console.log(html_beautify(wrapper.html()));
    expect(wrapper.html()).toMatchSnapshot();
  });
});
