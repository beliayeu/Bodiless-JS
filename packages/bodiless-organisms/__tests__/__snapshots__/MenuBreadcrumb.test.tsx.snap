// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`asBreadcrumbsClean creates breadcrumbs for basic 1-level menu 1`] = `"<ul><li data-bl-activator=\\"true\\"><a href=\\"/products\\"></a></li></ul>"`;

exports[`asBreadcrumbsClean creates breadcrumbs for basic 2-level menu 1`] = `"<ul><li><a href=\\"/products\\"></a></li><span></span><li><a href=\\"/products/productA\\"></a></li></ul>"`;
